Query improvements

lukedirtwalker opened the pull request:
More fixes, and implementation of a Join query.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44634312%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44634978%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44635018%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44635374%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44635680%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44635848%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44636028%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44636277%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44636345%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44637199%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23issuecomment-156051872%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23issuecomment-156051872%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Cool.%20Thanks.%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A48%3A52Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%2075e08d8ebcdb3d37bebd40222b4c5bb58c74bb18%20InformationScripting/src/queries/Join.cpp%2084%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44635018%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20don%27t%20understand%20the%20problem.%20This%20is%20provided%20by%20the%20user%2C%20so%20what%27s%20the%20issue%3F%20Why%20would%20we%20need%20more%20flexibility%20%28example%29%3F%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A14%3A33Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22This%20and%20the%20next%20line%20belong%20to%20line%2089%20just%20below.%20And%20in%20fact%20I%20just%20realized%20that%20it%20is%20plain%20wrong.%20I%20thought%20we%20had%20to%20extract%20the%20Property%20value%20to%20do%20the%20comparison%2C%20but%20we%20can%20just%20use%20Property%3A%3Aoperator%3D%3D%20so%20it%20will%20work%20for%20any%20property%20value.%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A18%3A04Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed%20in%20a%20new%20commit.%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A22%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/Join.cpp%3AL1-156%22%7D%2C%20%22Pull%2075e08d8ebcdb3d37bebd40222b4c5bb58c74bb18%20InformationScripting/src/dataformat/Tuple.cpp%2018%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/211%23discussion_r44634312%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22So%20this%20is%20nice%2C%20but%20do%20you%20still%20think%20we%20should%20support%20the%20old%20behavior%20%28not%20providing%20a%20tag%20and%20using%20the%20tag%20of%20the%20first%20part%20of%20the%20tuple%29%3F%20Are%20there%20any%20drawbacks%20to%20removing%20the%20old%20behavior%3F%20What%20are%20the%20advantages%20of%20keeping%20this%20flexibility%3F%20I%20feel%20like%20it%20might%20cause%20confusion.%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A05%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Well%20I%20left%20the%20old%20behavior%20mostly%20that%20it%20is%20not%20a%20too%20invasive%20change.%20For%20example%20for%20ast%20nodes%20it%20makes%20sense%2C%20but%20we%20can%20also%20change%20it%20to%20ast.node%20or%20something.%20There%20are%20no%20drawbacks%20of%20removing%20the%20old%20behavior.%20I%20think%20we%20can%20change%20it%2C%20if%20you%27d%20like%20so.%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A13%3A55Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22Well%2C%20my%20only%20concern%20is%20that%20having%20this%20flexibility%20slightly%20complicates%20our%20basic%20data%20model.%20But%20I%20am%20not%20sure%20if%20that%27s%20indeed%20a%20problem%20or%20not.%5Cr%5Cn%5Cr%5CnPerhaps%20a%20slightly%20nicer%20solution%20is%20to%20say%3A%5Cr%5Cn-%20A%20tuple%20always%20has%20a%20tag%5Cr%5Cn-%20You%20can%20either%20provided%20manually%5Cr%5Cn-%20Or%20else%20it%20gets%20automatically%20assgined%20at%20construction%20from%20the%20tag%20of%20the%20first%20tuple%20element.%20This%20is%20very%20similar%20to%20the%20current%20solution%20with%20two%20differences%3A%5Cr%5Cn%20-%20In%20the%20default%20case%2C%20we%20insist%20that%20the%20tuple%20is%20initialized%20with%20at%20least%20one%20value%2C%20so%20that%20the%20empty%20case%20at%20the%20bottom%20of%20this%20function%20falls%20out.%5Cr%5Cn%20-%20We%20change%20this%20method%20to%20always%20return%20%60tag_%60%20and%20inline%20it%2C%20and%20move%20the%20entire%20initialization%20to%20the%20initializer%20list%20in%20the%20constructors.%20This%20is%20a%20bit%20more%20explicit.%5Cr%5Cn%5Cr%5CnWhat%20do%20you%20think%3F%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A24%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Sounds%20good%20to%20me.%5Cr%5CnShould%20I%20assert%20that%20the%20lists%20at%20construction%20are%20not%20empty%20or%20how%20should%20I%20handle%20this%20properly%3F%5Cr%5CnNote%20that%20this%20constructor%20can%20also%20be%20called%20from%20scripts%2C%20so%20an%20assertion%20might%20not%20be%20the%20optimal%20solution%3F%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A27%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20think%20an%20assertion%20is%20the%20right%20thing.%20This%20is%20not%20a%20user%20error%2C%20but%20a%20programming%20error%20in%20the%20sense%20that%20one%20should%20have%20provided%20either%20a%20tag%20or%20some%20initial%20values.%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A30%3A18Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22BTW%3A%20putting%20the%20tag%20of%20a%20tuple%20at%20the%20end%20when%20constructing%20a%20new%20one%20is%20akward.%20The%20tag%20should%20be%20the%20first%20argument.%20Please%20do%20it%20like%20that%20%28you%27ll%20need%20another%202%20constructors%29.%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A31%3A02Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Fixed%22%2C%20%22created_at%22%3A%20%222015-11-12T09%3A41%3A26Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/dataformat/Tuple.cpp%3AL28-45%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [ ] <a href='#crh-comment-Pull 75e08d8ebcdb3d37bebd40222b4c5bb58c74bb18 InformationScripting/src/dataformat/Tuple.cpp 18'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/211#discussion_r44634312'>File: InformationScripting/src/dataformat/Tuple.cpp:L28-45</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> So this is nice, but do you still think we should support the old behavior (not providing a tag and using the tag of the first part of the tuple)? Are there any drawbacks to removing the old behavior? What are the advantages of keeping this flexibility? I feel like it might cause confusion.
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Well I left the old behavior mostly that it is not a too invasive change. For example for ast nodes it makes sense, but we can also change it to ast.node or something. There are no drawbacks of removing the old behavior. I think we can change it, if you'd like so.
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Well, my only concern is that having this flexibility slightly complicates our basic data model. But I am not sure if that's indeed a problem or not.
  Perhaps a slightly nicer solution is to say:
- A tuple always has a tag
- You can either provided manually
- Or else it gets automatically assgined at construction from the tag of the first tuple element. This is very similar to the current solution with two differences:
- In the default case, we insist that the tuple is initialized with at least one value, so that the empty case at the bottom of this function falls out.
- We change this method to always return `tag_` and inline it, and move the entire initialization to the initializer list in the constructors. This is a bit more explicit.
  What do you think?
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Sounds good to me.
  Should I assert that the lists at construction are not empty or how should I handle this properly?
  Note that this constructor can also be called from scripts, so an assertion might not be the optimal solution?
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> I think an assertion is the right thing. This is not a user error, but a programming error in the sense that one should have provided either a tag or some initial values.
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> BTW: putting the tag of a tuple at the end when constructing a new one is akward. The tag should be the first argument. Please do it like that (you'll need another 2 constructors).
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Fixed
- [x] <a href='#crh-comment-Pull 75e08d8ebcdb3d37bebd40222b4c5bb58c74bb18 InformationScripting/src/queries/Join.cpp 84'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/211#discussion_r44635018'>File: InformationScripting/src/queries/Join.cpp:L1-156</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> I don't understand the problem. This is provided by the user, so what's the issue? Why would we need more flexibility (example)?
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> This and the next line belong to line 89 just below. And in fact I just realized that it is plain wrong. I thought we had to extract the Property value to do the comparison, but we can just use Property::operator== so it will work for any property value.
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Fixed in a new commit.
- [ ] <a href='#crh-comment-General Comment'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/211#issuecomment-156051872'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Cool. Thanks.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/211?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/211?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/211'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

lukedirtwalker committed:
Call builLookupHash before calling find on a tree.

This makes sure we don't hit an assertion when running the CHistory
command.

lukedirtwalker committed:
Fix a spelling mistake.

lukedirtwalker committed:
Fix the wrapping of the hash method of Tuple

The signature has to match the python signature.

lukedirtwalker committed:
Don't print space before commas in PythonWrapperGen

lukedirtwalker committed:
VC-Query: Add argument to only consider commits that affect a node

lukedirtwalker committed:
Make it possible to have custom tags in Tuple

For example for the VersionControlQuery it doesn't make sense to have
the first value as tag, we rather want to address it with change.id
instead of change.change

lukedirtwalker committed:
Add a join query

Example: changes -sc=affect -c=1|join -
v=change.id,commit.author,change.ast -o=commit.id=change.id -
as=astAuthor

dimitar-asenov commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/211'>Refresh</a>.<a href='#crh-update-none'></a>

dimitar-asenov commented on InformationScripting/src/dataformat/Tuple.cpp:
```diff
@@ -28,16 +28,17 @@
 
 QString Tuple::tag() const
 {
+	if (!tag_.isNull()) return tag_;
```
So this is nice, but do you still think we should support the old behavior (not providing a tag and using the tag of the first part of the tuple)? Are there any drawbacks to removing the old behavior? What are the advantages of keeping this flexibility? I feel like it might cause confusion.

lukedirtwalker replied:
Well I left the old behavior mostly that it is not a too invasive change. For example for ast nodes it makes sense, but we can also change it to ast.node or something. There are no drawbacks of removing the old behavior. I think we can change it, if you'd like so.

dimitar-asenov replied:
Well, my only concern is that having this flexibility slightly complicates our basic data model. But I am not sure if that's indeed a problem or not.

Perhaps a slightly nicer solution is to say:
- A tuple always has a tag
- You can either provided manually
- Or else it gets automatically assgined at construction from the tag of the first tuple element. This is very similar to the current solution with two differences:
  - In the default case, we insist that the tuple is initialized with at least one value, so that the empty case at the bottom of this function falls out.
  - We change this method to always return `tag_` and inline it, and move the entire initialization to the initializer list in the constructors. This is a bit more explicit.

What do you think?

lukedirtwalker replied:
Sounds good to me.
Should I assert that the lists at construction are not empty or how should I handle this properly?
Note that this constructor can also be called from scripts, so an assertion might not be the optimal solution?

dimitar-asenov replied:
I think an assertion is the right thing. This is not a user error, but a programming error in the sense that one should have provided either a tag or some initial values.

dimitar-asenov replied:
BTW: putting the tag of a tuple at the end when constructing a new one is akward. The tag should be the first argument. Please do it like that (you'll need another 2 constructors).

lukedirtwalker replied:
Fixed

dimitar-asenov commented on InformationScripting/src/queries/Join.cpp:
```diff
@@ -0,0 +1,155 @@
+
+	for (const auto& tuple1 : tag1Tuples)
+	{
+		// the id always has to be a string, or some other statically defined value. :(
```
I don't understand the problem. This is provided by the user, so what's the issue? Why would we need more flexibility (example)?

lukedirtwalker replied:
This and the next line belong to line 89 just below. And in fact I just realized that it is plain wrong. I thought we had to extract the Property value to do the comparison, but we can just use Property::operator== so it will work for any property value.

lukedirtwalker replied:
Fixed in a new commit.

lukedirtwalker committed:
Fix Join to work for any type of the id value.

lukedirtwalker committed:
Require a tag in Tuple

It is either the name of the first namedproperty (implicit) or an
explicit value passed to the constructor.

dimitar-asenov commented:
Cool. Thanks.
